File: HW2/hw2.py
import sys
import random
import math

def dotproduct(u , v):
    assert len(u) == len(v), "dotproduct: u and v must be of same length"
    dp = 0
    for i in range(0,len(u), 1):
        dp += u[i] * v[i]
    return dp


datafile = sys.argv[1]
f = open(datafile)
data = []
i = 0
l = f.readline()

##############
#Read Data#
##############
while(l != ''):
    a = l.split()
    l2 = []
    for j in range(0, len(a), 1):
        l2.append(float(a[j]))
    l2.append(1)
    data.append(l2)
    l = f.readline()

rows = len(data)
cols = len(data[0])
f.close()

##############
#read labels#
##############

labelfile = sys.argv[2]
f = open(labelfile)
trainlabels = {}
n = []
n.append(0)
n.append(0)

l = f.readline()
while(l != ''):
    a = l.split()
    trainlabels[int(a[1])] = int(a[0])
    l = f.readline()
    n[int(a[0])] +=1


##################
##Initialize w
##################
w = []
for i in range(0,cols,1):
    w.append(random.uniform(-0.01,0.01))

#########################################################
#########################################################
#########################################################
#########################################################

##################
#Gradient Descent#
##################

###Initializations

# eta = 0.001
#for breast cancer
#eta = 0.000000001
#for climate
#eta = 0.001
#for qsar
#eta = 0.000001
#for ionosphere
eta = 0.0001
#for hill valley
#eta = 0.001

dellf = []
for j in range(0, cols, 1 ):
    dellf.append(0)

prevobj = 100000000
obj = prevobj - 10

###########Main Iteration Loop############
# while(abs(prevobj - obj)> 0):
# while(prevobj - obj> 0.001):
# while(prevobj - obj> 0):
while(abs(prevobj - obj)> 0.001):
# while(abs(prevobj - obj)> 0.00001):
# while(abs(prevobj - obj)> 0.000000001):
# while(abs(prevobj - obj)> 0 and iter < 10000):
# while(iter < 1):
    prevobj = obj
    ################ Reset dellf to 0 #################
    for j in range(0,cols,1):
        dellf[j] = 0
    
    ###############Compute dellf ################
    for i in range(0,rows,1):
        if(trainlabels.get(i) != None):
            dp = dotproduct(w,data[i])
            for j in range(0,cols,1):
                dellf[j]+= (trainlabels.get(i) - dp)*data[i][j]

    ##########Update w ####################
    for j in range(0, cols, 1):
        w[j] += eta * dellf[j]
    
    error = 0

    ############ Compute error ###################
    for i in range(0,rows,1):
        if(trainlabels.get(i) != None):
            error += (trainlabels.get(i) - dotproduct(w,data[i]))**2
    obj = error
    print("Objective is ", error)


print(w)
wlen = math.sqrt(w[0]**2 + w[1]**2)
dist_to_origin = abs(w[2])/wlen
print("Distance to Origin", dist_to_origin)

wlen = 0
for i in range(0, len(w), 1):
    wlen += w[i]**2
wlen = math.sqrt(wlen)
print("wlen = ", wlen)                

#####################################
#####Classify Unlabbled Points########
####################################
for i in range(0,rows,1):
    if(trainlabels.get(i) == None):
    # if(trainlabels.get(i) != None):
    #if(1):
        dp = dotproduct(w,data[i])
        if(dp<0):
            print("0 ", i)
        else:
            print("1 ", i)

File: HW2/hw2.pl
# pyth.RunPerl.ext;

$data = shift;
$trainlabels = shift;

#############
##Read data and labels from files
#############
open(IN, $data);
@data = ();
$i = 0;
while (eof(IN) != 1){
    $l = <IN>;
    @a=split(/\s+/,$l);
    for(my $j=0; $j<@a; $j++){
        $data[$i][$j]=$a[$j];
    }
    $data[$i][scalar(@a)]=1;
    $i++;
}
close(IN);
$rows = scalar(@data);
$ref = $data[0];
$cols = scalar(@$ref);

open(IN, $trainlabels);
@class = ();
@class_size = ();
while( eof(IN) != 1 ){
    $l=<IN>;
    @a=split(/\s+/,$l);
    $class[$a[1]] =$a[0];
    $class_size[$a[0]] = $class_size[$a[0]] +1;
    if($class[$a[1]] == 0) {
        $class[$a[1]] = -1;
    }
}
#############
##Initialize w
#############

@w = ();
for(my $j=0; $j<$cols; $j++){
    $w[$j] = .02 * rand(1) - 0.01;
    # $w[$j] = .0002 * rand(1) - 0.0001;

}

#############
##Gradient descent iteration
#############
$eta = .001;
# $eta = .0001;
# $eta = .000001;
# $eta = .000000001;

# $prevobj = 100000000;
# $obj = $prevobj - 10;

#while(abs($prevobj - $obj)> 0.001){
# $prevobj = $obj;
for(my $k=0; $k<100000; $k++){
    ####compute dellf###
    @dellf = ();
    for(my $i=0; $i<$rows; $i++){
        if(defined($class[$i])){
            $dp = &dot_product(\@w, $data[$i]);
            for(my $j = 0; $j<$cols; $j++){
                $dellf[$j] += ($class[$i]-$dp)*$data[$i][$j];
            }
        }
    }

#######################################
############## Update W ###############
#######################################
    for(my $j=0;  $j<$cols; $j++){
        $w[$j] = $w[$j] + $eta*$dellf[$j];
    }

    $error = 0;
   
    $obj = ();
    
#######################################
###########Compute error###############
#######################################
    for(my $i=0; $i < $rows; $i++){
        
        if(defined($class[$i])){
            
            $error += ($class[$i] - &dot_product(\@w, $data[$i]))**2;
            push @obj, $error;
            
        }
    }
    # $obj = $error;
    print "error = $error\n";
    # print "array = @obj\n";
    
}
# }


print "w = ";
$normw = 0;
for(my $j=0; $j<$cols-1; $j++){
    $normw += $w[$j]**2;
    print "$w[$j] ";
}
print "\n";
$normw = sqrt($normw);
print "||w||=$normw\n";
$d_origin = $w[scalar(@w)-1]/$normw;
print "distance to origin = $d_origin\n";
@abs_d_origin = abs($d_origin);
print "abs(w0/||w||)  = @abs_d_origin\n";

################################
#Prediction
#############################
open(OUT, ">p_output");
for (my $i = 0; $i < $rows; $i++){
    if(!defined($class[$i])){
        $dp = &dot_product(\@w, $data[$i]);
        if($dp>0){
            print OUT "1 $i\n";
        }
        else{
            print OUT "0 $i\n";
            }
    }
}
close(OUT);

#####################################
###Sub Routine
##########################
sub dot_product{

    my $refw = $_[0];
    my $refx = $_[1];
    my $dp = 0;
    for(my $j=0; $j<$cols; $j++){
        $dp += $$refw[$j] * $$refx[$j];
    }
    return $dp;
}